### ISSUE TITLE
QUESTION: Is it possible to run Nginx server simultaneously to UH-vpn server?

### ISSUE THREAD

**Issue by sunny75016** [ultrahorizon/UH-VPN-Docs#43]
Hope you both are safe during the COVID19 crisis.

If possible I would like to run the nginx (or apache2) server on port 443 at the same time as uh-vpn server. So where the server name is abc.def.ghi the traffic should be handled by nginx. Otherwise, the traffic should be handled by uh-vpn as done now.

Technically, is this possible? Would you like to include some instructions? This for considering this as a feature request.

**Comment by jwsi** [comment]
I think a nice way to do this is to split based on transport protocol:

- Run nginx on TCP 80, 443 etc... this will handle all HTTP, HTTPS traffic and so on.
- Run UH VPN Server on UDP 443.

This way all UDP traffic is handled via UH VPN, and TCP traffic by nginx. Would that be suitable for your use case?

In general it is advisable to avoid using TCP based VPN servers as they'll be less performant than their UDP counterparts as needless re-transmissions will occur with TCP based tunnels.

**Comment by sunny75016** [comment]
Thanks @jwsi - I am already running UH VPN on both TCP/443 and UDP/443 on the same DigitalOcean droplet; because some environments block UDP. So the challenge is now is to split the TCP/443 traffic into two parts - That directed towards the normal nginx website and the remaining towards UH-VPN.   I have seen a similar split being done for Jitsi where nginx serves as a reverse proxy; however, that was only possible as you can select which port jitsi and nginx run on (i.e. 4443 and 443 respectively). Unfortunately, in the case of UH-VPN - as the port is pre-set on both the server end and on the client end - we need to figure a way to achieve this split.

**Comment by jwsi** [comment]
Hey @sunny75016,

I see your predicament. The main issue at play is that they'll be a socket conflict as both UH VPN Server and nginx will try to listen on `tcp *:443` and the underlying Linux kernel will throw a `socket already in use` error to one of the applications.

**Easy solution**
The only reason we suggest to run UH VPN Server on UDP 443 and TCP 443 is attempt to emulate the QUIC and HTTPS protocols for DPI purposes. Therefore, it's possible to run UH VPN Server and achieve the same DPI bypass features as the TCP 443 solution by setting UH VPN Server to run on TCP 8443 (the alternative HTTPS port). This will then free up TCP 443 entirely for nginx. This in my opinion is the nicest and simplest solution as it retains all the existing functionality. I run my TCP VPN server on TCP 8443 to free up TCP 443 for web applications.

**Hard way**
You could set UH VPN Server to run on some random TCP port, say 1234. Then you can have nginx running on TCP 443. Then you could examine the host header (url) of requests to determine whether it's web traffic or not (or do some DPI yourself). If a host header is present and is the url you expect then you route to a web app, otherwise route to the UH VPN Server. This is a bit of a hacky solution and I think you may enter a few issues especially when HTTPS comes into play as nginx will likely attempt to take over the cryptographic handshake that is meant for the VPN. Therefore, at this point, for reliability and performance I'd say that if TCP 443 is very important to you, then it's probably best to make a separate VPN and web droplet, otherwise switch to TCP 8443 which has no real drawbacks at all.

**tldr; Set UH VPN Server to run on UDP 443 and TCP 8443**
